src/compiler/emitter.ts

```diff
@@ -1343,6 +1343,35 @@ const enum PipelinePhase {
     Emit,
 }
 
+const printerCache = new Map<string, Printer>();
+
+// If updating this, make sure the cache key is updated too.
+/** @internal */
+export type ReusablePrinterOptions = Pick<PrinterOptions, "removeComments" | "neverAsciiEscape" | "omitTrailingSemicolon" | "module" | "target" | "newLine">;
+
+/** @internal */
+export function createOrReusePrinter(o: ReusablePrinterOptions = {}) {
+    const key = keyBool(o.removeComments)
+        + keyBool(o.neverAsciiEscape)
+        + keyBool(o.omitTrailingSemicolon)
+        + keyNum(o.module)
+        + keyNum(o.target)
+        + keyNum(o.newLine);

```

Comment:
Believe it or not, this is the fastest cache key of the variants I tried.

Reply:
It's actually faster than using a simple native template string like
```ts
`${o.removeComments}|${o.neverAsciiEscape}|${o.omitTrailingSemicolon}|${o.module}|${o.target}|${o.newLine}`
```
? Feels odd.

Reply:
Oh, but, perhaps more relevant - we have a `getKeyForCompilerOptions` function already. It's used by the `documentRegistry` in the language service. If there is actually a big difference in time spent making the key, maybe some changes to it would be warranted as well? I'd certainly prefer it if we reused the existing option-key-generating-mechanism if we could.

Reply:
So, two things:

- Writing out `keyBool(...) + "|" + keyBool(...) + "|" + ...` was just as fast as writing out `${keyBool(...)}|${keyBool(...)}...`.
- Moving the separators into the "keys" (i.e. how I did `u,` `1,` `0,`) instead of putting them the template was _twice_ as fast.

So, I opted for the variant which doesn't make the line go way way to the right as the number of parts increases.

Reply:
That being said it's still like millions of iterations per second so unlikely to matter; I was really sad to find that `[..., ..., ...].join("|")` was an order of magnitude slower than these methods.

I'll go take a look at that other function.

Reply:
Ah, `getKeyForCompilerOptions` is quite literally what I said was slow above; given the arbitrary nature of the big list of compiler keys, I'm not sure I could change that.

The checker itself does have a few internal caches which do the hardcoded thing you said (and I did first), hard to say if they'd really benefit.

I'm happy to make this whatever makes sense, I just knew for this one I had to combine like 6 things and the length was getting unwieldy (so I resorted to helpers and multiple lines).

Reply:
All 2 existing usages of `getKeyForCompilerOptions` actually use a fixed list of keys - longer lists, to be sure, made by filtering the full list of options by a boolean flag set on the option, but still fixed lists. In module resolution, we actually use these keys quite a lot - if it's bad enough to matter here, I'd think it'd matter there, too.

> Writing out keyBool(...) + "|" + keyBool(...) + "|" + ... was just as fast as writing out ${keyBool(...)}|${keyBool(...)}....

As for the simpler case, my focus on moreso why the coercion functions were needed at all, and a simple string concatenation wouldn't suffice. Since the order is fixed, it's not like you need to distringuish between strings and `undefined` or `true` or `false` or the like - is there a perf benefit to writing `value === undefined ? "u," : value ? "1," : "0,"` over `${value},`?

Reply:
Ah, I see. I tested that one out and it was middling in performance.

https://jsbench.me/prldbzl6dj/2 is what I was using to test, but I also can just revert these last few commits and go back to the version which hardcodes the 4 most common ones.

Reply:
For your future cache key micro-optimizing endeavors: If string concatenation is the slow bit, seems like it's best to skip on strings almost entirely:
```ts
function bitwise(o) {
	return ""+((o.removeComments ? 1 : 0)
        | (o.neverAsciiEscape ? 2 : 0)
        | (o.omitTrailingSemicolon ? 4 : 0)
        | (o.module << 4) // 0-199
        | (o.target << 12) // 0-100
        | (o.newLine << 20)); // 0-1
}
```
is ~5% faster than the string-based next best 😛

Reply:
That's a good idea (doesn't even have to be a string key either, so we could skip the ""+)